### PR TITLE
FastIntersectionFinder

### PR DESCRIPTION
**Pull request by PyMarc2**
In the case of a scene with a `SpacePartition`, it is much faster to traverse the tree structure of the partition to find the intersection.

The algorithm is a recursive backtrack algorithm. Search starts at ray's origin node and backtracks it's way up the tree until all potential node are declared "visited".

### CODE REVIEW

File: pytissueoptics/scene/tree/spacePartition.py [hunk 1]
```diff
@@ -40,24 +41,33 @@ def searchPoint(self, point: Vector, node: Node = None) -> Node:
...
                 return node
 
-        return self.searchPoint(point, isInside)
+            else:
+                return None
+
+        return isInside
+
+    def findLeafNodesContainingPolygon(self, polygon):
+        pass
```
Comment (JLBegin) [inline pytissueoptics/scene/tree/spacePartition.py:57]: What's up with that ?

File: pytissueoptics/scene/tree/spacePartition.py [hunk 2]
```diff
@@ -105,3 +115,24 @@ def getLeafBoundingBoxesAsCuboids(self) -> List[Cuboid]:
...
+            node = self._root
+            jsonFile.append({})
+            jsonFile[0]["depth"] = node.depth
+            jsonFile[0]["size"] = len(node.polygons)
+            jsonFile[0]["bbox"] = f"{node.bbox:.2f}"
+            jsonFile[0]["children"] = []
+            self.getJSONBranching(node, jsonFile[0]["children"])
+
+        if node.isRoot:
+            return json.dumps(jsonFile)
```
Comment (JLBegin) [inline pytissueoptics/scene/tree/spacePartition.py:138]: I understand this is for debugging. 
If you really want to keep it, put it private or outside package maybe.

File: pytissueoptics/scene/tree/treeConstructor/binary/binaryNodeSplitter.py [hunk 3]
```diff
@@ -82,3 +82,56 @@ def _searchMinSAH(self, nodeBbox, polygons, splitAxis, aMin, step):
...
+
+        if minSAH + splitCost < nodeSAH:
+            stopCondition = False
+
+        else:
+            stopCondition = True
+
+        polygonGroups = self._polygonCounter.count(splitLine, splitAxis, polygons)
+        groupBbox = self._getNewChildrenBbox(newBounds, splitAxis, splitLine)
+        return SplitNodeResult(stopCondition, splitAxis, splitLine, groupBbox, polygonGroups)
```
Comment (JLBegin) [inline pytissueoptics/scene/tree/treeConstructor/binary/binaryNodeSplitter.py:115]: Duplicated code at HardSAH.
Maybe create an SAH abstract class or something. Or inherit from HardSAH if this is a variation of the latter.

File: pytissueoptics/scene/tree/treeConstructor/binary/binaryNodeSplitter.py [hunk 4]
```diff
@@ -82,3 +82,56 @@ def _searchMinSAH(self, nodeBbox, polygons, splitAxis, aMin, step):
...
+            tempRightBbox.update(splitAxis, "min", split)
+            newSAH = len(left) * tempLeftBbox.getArea() + len(right) * tempRightBbox.getArea()
+
+            if i == 1:
+                minSAH = newSAH
+
+            if newSAH < minSAH:
+                minSAH = newSAH
+                lowestIndexSAH = i
+
```
Comment (JLBegin) [inline pytissueoptics/scene/tree/treeConstructor/binary/binaryNodeSplitter.py:135]: Duplicated code at HardSAH.

File: pytissueoptics/scene/tree/treeConstructor/utils.py [hunk 5]
```diff
@@ -14,3 +14,9 @@ def meanCentroid(axis: str, polygons: List[Polygon]):
             average += polygon.centroid.z
     average = average / len(polygons)
     return average
+
+def getPolygonsBbox(polygons: List[Polygon]):
+    bbox = BoundingBox([0, 0], [0, 0], [0, 0])
+    for polygon in polygons:
+        bbox.extendTo(polygon.bbox)
+    return bbox
```
Comment (JLBegin) [inline pytissueoptics/scene/tree/treeConstructor/utils.py:22]: If these 2 private methods were moved out to utils, it's to test them, but they are currently not tested.

File: pytissueoptics/scene/logger/__init__.py [hunk 6]
```diff
@@ -1 +1 @@
-from .logger import Logger
+from .logger import Logger, Segment
```
Comment (JLBegin) [inline pytissueoptics/scene/logger/__init__.py:1]: This Segment import is not used.

File: pytissueoptics/scene/intersection/intersectionFinder.py [hunk 7]
```diff
@@ -53,25 +77,53 @@ def _findBBoxIntersectingSolids(self, ray) -> Union[List[Tuple[float, Solid]], N
...
+                    if bboxIntersection is not None:
+                        if closestIntersection is not None:
+                            if (bboxIntersection - ray.origin).getNorm() > closestIntersection.distance:
+                                continue
+                        intersection = self._exploreNodeForIntersection(ray, child)
+                        if intersection is not None:
+                            if closestIntersection is None:
+                                closestIntersection = intersection
+                            elif intersection.distance < closestIntersection.distance:
+                                closestIntersection = intersection
```
Comment (JLBegin) [inline pytissueoptics/scene/intersection/intersectionFinder.py:123]: Not sure what this elif is for, but it is not covered by the tests.

File: pytissueoptics/scene/tree/node.py [hunk 8]
```diff
@@ -11,6 +11,7 @@ def __init__(self, parent: 'Node' = None, polygons: List[Polygon] = None, bbox:
         self._polygons = polygons
         self._bbox = bbox
         self._depth = depth
+        self._visited = False
 
     @property
     def parent(self):
```
Comment (JLBegin) [inline pytissueoptics/scene/tree/node.py:17]: Not used.

File: pytissueoptics/scene/tree/treeConstructor/binary/binaryNodeSplitter.py [hunk 9]
```diff
@@ -82,3 +82,56 @@ def _searchMinSAH(self, nodeBbox, polygons, splitAxis, aMin, step):
 
         splitLine = aMin + lowestIndexSAH * step
         return splitLine, minSAH
+
+
+class ShrankBoxSAHNodeSplitter(NodeSplitter):
```
Comment (JLBegin) [inline pytissueoptics/scene/tree/treeConstructor/binary/binaryNodeSplitter.py:87]: Not tested.

**Review by JLBegin — CHANGES_REQUESTED** [review]
Good design and seems to work well. 
I will check tests and performance after feedback is addressed.

File: pytissueoptics/scene/geometry/bbox.py [hunk 10]
```diff
@@ -22,6 +22,15 @@ def __eq__(self, other: 'BoundingBox') -> bool:
...
 
+    def __format__(self, formatSpec):
+        f_xMin = float(format(self.xMin, formatSpec))
+        f_xMax = float(format(self.xMax, formatSpec))
+        f_yMin = float(format(self.yMin, formatSpec))
+        f_yMax = float(format(self.yMax, formatSpec))
+        f_zMin = float(format(self.zMin, formatSpec))
+        f_zMax = float(format(self.zMax, formatSpec))
+        return str([[f_xMin, f_xMax], [f_yMin, f_yMax], [f_zMin, f_zMax]])
+
```
[the remark this answers is not in the document]
Comment (PyMarc2) [reply pytissueoptics/scene/geometry/bbox.py:33]: It is analog to the `__repr__`, but it allows to use format strings, in instance when bbox is printed, but one would want to round to the 2nd decimal f"{object.bbox:.2f}" would call `__format__` instead of `__repr__`.

File: pytissueoptics/scene/examples/example4.py [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+    loader = Loader()
+    scene = loader.load("exampleFile.obj")
+
+    viewer = MayaviViewer()
+    viewer.add(*scene, representation="surface")
+    viewer.show()
+
+
+if __name__ == "__main__":
+    exampleCode()
```
[the remark this answers is not in the document]
Comment (PyMarc2) [reply pytissueoptics/scene/examples/example4.py]: resolved with commit: 4c6fe8bedf206e4e3fcaf44ca3c258b420c14d38

File: pytissueoptics/scene/geometry/bbox.py [hunk 12]
```diff
@@ -140,6 +149,20 @@ def extendTo(self, other: 'BoundingBox'):
...
+            self._xLim[1] = other.xMax
+        if other.yMin > self.yMin:
+            self._yLim[0] = other.yMin
+        if other.yMax < self.yMax:
+            self._yLim[1] = other.yMax
+        if other.zMin > self.zMin:
+            self._zLim[0] = other.zMin
+        if other.zMax < self.zMax:
+            self._zLim[1] = other.zMax
+
```
[the remark this answers is not in the document]
Comment (PyMarc2) [reply pytissueoptics/scene/geometry/bbox.py:165]: resolved with commit: d68434e4251a451c068485825d7a28affbb21507

File: pytissueoptics/scene/intersection/intersectionFinder.py (hunk 7 above)
Comment (PyMarc2) [reply pytissueoptics/scene/intersection/intersectionFinder.py:123]: In the event that a new `bbox` is hit and that the hit was closer than the last closest surface, we must verify that the true intersection is truly closer. It might be the case that the `bbox` is hit, but the actual interface is further away. (Not our case actually because `bbox` cannot touch each other, but it is a necessary verication that would have been forgotten in the future.

File: pytissueoptics/scene/intersection/intersectionFinder.py [hunk 13]
```diff
@@ -53,25 +77,53 @@ def _findBBoxIntersectingSolids(self, ray) -> Union[List[Tuple[float, Solid]], N
...
+
```
[the remark this answers is not in the document]
Comment (PyMarc2) [reply pytissueoptics/scene/intersection/intersectionFinder.py:124]: I did refactor down at some point but reverted. There is a complication in the recursive algorithm I cannot pin point.

[the remark this answers is not in the document]
Comment (PyMarc2) [reply pytissueoptics/scene/intersection/intersectionFinder.py]: stairway resolved with commit: 7c779d63b70f2a9b23ea0f44baec8f24c8bd3e6e

File: pytissueoptics/scene/tests/intersection/testIntersectionFinder.py [hunk 14]
```diff
@@ -4,10 +4,14 @@
...
-from pytissueoptics.scene.intersection import SimpleIntersectionFinder, Ray
+from pytissueoptics.scene.tree import SpacePartition
+from pytissueoptics.scene.tree.treeConstructor.binary import SAHWideAxisTreeConstructor
+from pytissueoptics.scene.scene import Scene
+from pytissueoptics.scene.intersection import SimpleIntersectionFinder, FastIntersectionFinder, Ray
 
 
-class TestAnyIntersectionFinder:
+class TestAnyIntersectionFinder(unittest.TestCase):
+
```
[the remark this answers is not in the document]
Comment (PyMarc2) [reply pytissueoptics/scene/tests/intersection/testIntersectionFinder.py]: resolved with commit 277facb67aeab4e00f9a260d86afe077e89f522d

File: pytissueoptics/scene/tree/spacePartition.py (hunk 1 above)
Comment (PyMarc2) [reply pytissueoptics/scene/tree/spacePartition.py]: resolved with commit e9e8345717724422e1e353443aef9206531c3fd0

File: pytissueoptics/scene/tree/spacePartition.py (hunk 2 above)
Comment (PyMarc2) [reply pytissueoptics/scene/tree/spacePartition.py:135]: did put it outside the package. resolved with commit 61f6dd15c9719c1995e0c7f4bad10dc3969e3686

File: pytissueoptics/scene/tree/node.py (hunk 8 above)
Comment (PyMarc2) [reply pytissueoptics/scene/tree/node.py:17]: details needed.

Comment (JLBegin) [reply pytissueoptics/scene/tree/node.py:17]: public node.parent is never called in the code base

File: pytissueoptics/scene/logger/__init__.py (hunk 6 above)
Comment (PyMarc2) [reply pytissueoptics/scene/logger/__init__.py]: Resolved with 4408122614e7a8e265f5b5f95ee0644c2040a1cd

File: pytissueoptics/scene/tree/node.py (hunk 8 above)
Comment (PyMarc2) [reply pytissueoptics/scene/tree/node.py]: Resolved with 4408122614e7a8e265f5b5f95ee0644c2040a1cd

File: pytissueoptics/scene/tree/treeConstructor/utils.py (hunk 5 above)
Comment (PyMarc2) [reply pytissueoptics/scene/tree/treeConstructor/utils.py:22]: Resolved with 4408122614e7a8e265f5b5f95ee0644c2040a1cd

File: pytissueoptics/scene/tree/treeConstructor/binary/binaryNodeSplitter.py (hunk 4 above)
Comment (PyMarc2) [reply pytissueoptics/scene/tree/treeConstructor/binary/binaryNodeSplitter.py]: resolved with commit fd6db356bfb184947a23962841f3a516446f088c

File: pytissueoptics/scene/tree/treeConstructor/binary/binaryNodeSplitter.py (hunk 3 above)
Comment (PyMarc2) [reply pytissueoptics/scene/tree/treeConstructor/binary/binaryNodeSplitter.py]: resolved with commit fd6db356bfb184947a23962841f3a516446f088c

File: pytissueoptics/scene/scene/scene.py [hunk 15]
```diff
@@ -7,18 +7,14 @@
 
 
 class Scene:
-    def __init__(self, solids: List[Solid] = None, ignoreIntersections=False):
+    def __init__(self, solids: List[Solid] = None, ignoreIntersections=False, partitionScene=True):
         self._solids = []
```
Comment (JLBegin) [inline pytissueoptics/scene/scene/scene.py:11]: This is not used...

**Review by JLBegin — CHANGES_REQUESTED** [review]
- [x] (resolved) I think I would change IntersectionFinder init to accept only a Scene instance. All IntersectionFinders will share the same default constructor, but FastIntersectionFinder will have additional optionnal parameters to setup the SpacePartionning. Then FastIntersectionFinder will create the SpacePartition for the scene. We don't want the user to have to deal with space partitionning. And it should not be created inside the Scene itself because it's really only useful for FastIntersectionFinder.

File: pytissueoptics/scene/intersection/intersectionFinder.py [hunk 16]
```diff
@@ -53,25 +79,61 @@ def _findBBoxIntersectingSolids(self, ray) -> Union[List[Tuple[float, Solid]], N
...
+        # rayStartingNode = self._partition.searchPoint(ray.origin)
+        # if rayStartingNode is None:
+        #     if not self._boxIntersect.getIntersection(ray, self._partition.root.bbox):
+        #         return None
+        rayStartingNode = self._partition.root
+        intersection = self._exploreNodeForIntersection(ray, rayStartingNode)
+        return intersection
+
+    def _exploreNodeForIntersection(self, ray: Ray, node: Node, closestIntersection=None) -> Optional[Intersection]:
```
Comment (JLBegin) [inline pytissueoptics/scene/intersection/intersectionFinder.py:110]: I suggest the following structure: 

```python
    def findIntersection(self, ray: Ray) -> Optional[Intersection]:
        """
        ...
        """

        intersection = self._findIntersection(ray, self._partition.root)
        return intersection

    def _findIntersection(self, ray: Ray, node: Node, closestDistance=sys.maxsize) -> Optional[Intersection]:
        if node.isLeaf:
            intersection = self._findClosestPolygonIntersection(ray, node.polygons)
            return intersection

        if not self._nodeIsWorthExploring(ray, node, closestDistance):
            return None

        closestIntersection = None
        for child in node.children:
            intersection = self._findIntersection(ray, child, closestDistance)
            if intersection is None:
                continue
            if intersection.distance < closestDistance:
                closestDistance = intersection.distance
                closestIntersection = intersection

        return closestIntersection

    def _nodeIsWorthExploring(self, ray, node, closestDistance) -> bool:
        bboxIntersection = self._boxIntersect.getIntersection(ray, node.bbox)
        if bboxIntersection is None:
            return False
        bboxDistance = (bboxIntersection - ray.origin).getNorm()
        if bboxDistance > closestDistance:
            return False
        return True

```

File: pytissueoptics/scene/geometry/polygon.py [hunk 17]
```diff
@@ -27,6 +27,12 @@ def __init__(self, vertices: List[Vector], normal: Vector = None,
         self.resetCentroid()
         self.resetBoundingBox()
 
+    def __eq__(self, other: 'Polygon'):
+        for vertex in self._vertices:
+            if vertex not in other.vertices:
+                return False
+        return True
+
```
Comment (PyMarc2) [inline pytissueoptics/scene/geometry/polygon.py:35]: tests have been written. d7a115525df70bb577285f959be4065572570477

File: pytissueoptics/scene/geometry/bbox.py [hunk 18]
```diff
@@ -36,6 +45,16 @@ def fromVertices(cls, vertices: List[Vector]) -> 'BoundingBox':
...
+    @classmethod
+    def fromPolygons(cls, polygons: List['Polygon']) -> 'BoundingBox':
+        bbox = None
+        for polygon in polygons:
+            if bbox is not None:
+                bbox.extendTo(polygon.bbox)
+            else:
+                bbox = polygon.bbox.copy()
+        return bbox
+
```
Comment (PyMarc2) [inline pytissueoptics/scene/geometry/bbox.py:57]: test not written

Comment (PyMarc2) [reply pytissueoptics/scene/geometry/bbox.py:57]: resolved with  f32d470b280e1e3c7d6a3d1dba35dc784f459bb2

**Comment by PyMarc2** [comment]
@JLBegin

 ## PR is ready to review.
_Hopefully last review_

In the end, the most important part is the `FastBinaryTreeConstructor`, which doesn't use the same internal structure as all the other methods, but still has the same interface. It is a custom /  modified version of what is proposed in
`Wald, Ingo, and Vlastimil Havran. 2006. “On Building Fast Kd-Trees for Ray Tracing, and on Doing That in O(N Log N).” In 2006 IEEE Symposium on Interactive Ray Tracing, 61–69.`

`FastBinaryTreeConstructor` is now the default `TreeConstructor` for the `FastIntersectionFinder`.
The performance is equivalent to `SimpleIntersectionFinder` in low-polygon `Scene`.
In high polygon `Scene`, it does much better than all the other algorithms.
e.g: On a `Scene` with 10k poly, it is 100x faster than `SimpleIntersectionFinder`

There are lots of tests with edge cases for splitting. Extensive "unit" tests have not been done with all the scene, as they become quite complicated. Instead, the `profileIntersectionFinder.py` helps with visualizing the splitting data and comparing speed.

As of now, 3 tests do not pass, they are old tests which I am now caring for, but they are tedious and require manual computations, so until I resolve those, the PR is still ready to review.

**Review by JLBegin — CHANGES_REQUESTED** [review]
Some tests are too slow inside testIntersectionFinder.py (64% of test time). 

As discussed, there is still some debugging to do since only the shrank box constructor as the expected behavior on E2E tests.

File: pytissueoptics/scene/tree/spacePartition.py [hunk 19]
```diff
@@ -40,24 +40,30 @@ def searchPoint(self, point: Vector, node: Node = None) -> Node:
...
+        return isInside
 
     @property
-    def maxDepth(self):
+    def maxAuthorizedDepth(self) -> int:
         return self._maxDepth
 
     @property
-    def minLeafSize(self):
+    def minAuthorizedLeafSize(self) -> int:
```
[the remark this answers is not in the document]
Comment (PyMarc2) [reply pytissueoptics/scene/tree/spacePartition.py:59]: used in profiler

File: pytissueoptics/scene/tests/intersection/testIntersectionFinder.py [hunk 20]
```diff
@@ -85,18 +88,39 @@ def testGivenRayIsIntersectingMultipleSolids_shouldReturnClosestIntersection(sel
...
+    def setUp(self) -> None:
+        scene = PhantomScene()
+        self.intersectionFinder = FastIntersectionFinder(scene)
+    
+    def testGivenRayTowardsBackWall(self):
+        origin = Vector(0, 4, 0)
+        direction = Vector(0, 0, -1)
+        ray = Ray(origin, direction)
+        intersection = self.intersectionFinder.findIntersection(ray)
+        print(intersection)
```
[the remark this answers is not in the document]
Comment (PyMarc2) [reply pytissueoptics/scene/tests/intersection/testIntersectionFinder.py:126]: It is unfinished indeed

**Comment by PyMarc2** [comment]
# Draft PR N°4521
## Are we arriving yet? @JLBegin 
In the end, I did extensive testing on the performance and parameter set for the polygon splitting tree constructor. It turns out, traversal performance is not enhanced significantly. At best, it matches non-splitting sah-based techniques. 

### Why chose a non-splitting technique?
The reason is that, contrary to previously believed, to intersection cost with a triangle is higher than anticipated.
The splitting algorithm always splits a triangle into 3 triangles, which makes 2 additional intersections, compared with just sharing both triangles, which results in only 1 additional intersection. Also, the construction time is significantly greater for the split algorithm, as it needs to find intersections with each splitter triangle. The real cost comes from the additional nodes it generates, which uses `bbox.copy()` (`copy.deepcopy()`) which account for more than 50% of the construction time for all algorithms. It is outrageous, however, I did not try many alternative methods, as they all failed to accomplish a true, non-referenced copy.

The only place that splitting polygons has an effect is when there are big polygons spanning across multiple nodes. For us, this will almost never be the case.

### Validation

For the validation and benchmark, I built `benchmarkIntersectionFinder`, which allows for easy algorithm validation using Ludo's lidar technique. A uniform cloud of rays is created and all the intersections must it a correct object. It is easily verifiable because we use the SimpleIntersectionFinder as the Reference, so we just compare it with that. Here's the validation data:
image
And the validation pointCloud:

### Benchmark
For the benchmark, I use the 2 `scenes`.
1. I use the same `Scene` as for the validation, but with completely random rays. This is because it is the slowest `Scene`, so it truly tests the capabilities of the tree.
2. A `Scene` with objects containing lots of polygons in a low number of objects. The SimpleIntersectionFinder performs well for high-object-count, low-polygons ratio, whereas FastIntersectionFinder works well for the opposite. The intersection performance boost is difficult function to find, as it depends on
- Object position
- Number of objects
- Shape of objects
- Number of polygons of objects.
However, we can still draw the performance curve for polygons only if we use a single sphere, from order 0 to 5:
image
image
We see the improvement in linear with the number of polygons, which is a significant improvement for high number of polygons. I always took 1.6seconds to propagate 100k rays, no matter the number of polygons. This is impressive to me.

### A typical day
Finally, here's the data of a typical benchmark session:
image

File: pytissueoptics/scene/geometry/bbox.py [hunk 21]
```diff
@@ -140,6 +163,28 @@ def extendTo(self, other: 'BoundingBox'):
...
+            self._zLim[1] = other.zMax
+
+    def extendBy(self, value: float):
+        self._xLim[0] -= value
+        self._xLim[1] += value
+        self._yLim[0] -= value
+        self._yLim[1] += value
+        self._zLim[0] -= value
+        self._zLim[1] += value
+
```
Comment (PyMarc2) [inline pytissueoptics/scene/geometry/bbox.py:187]: not tested

File: pytissueoptics/scene/geometry/surfaceCollection.py [hunk 22]
```diff
@@ -5,8 +5,10 @@
 
 
 class SurfaceCollection:
-    def __init__(self):
-        self._surfaces: Dict[str, List[Polygon]] = {}
+    def __init__(self, surfaces=None):
+        if surfaces is None:
+            surfaces = {}
+        self._surfaces: Dict[str, List[Polygon]] = surfaces
```
Comment (PyMarc2) [inline pytissueoptics/scene/geometry/surfaceCollection.py:12]: This is to allow custom solids in custom scenes

File: pytissueoptics/scene/intersection/raySource.py [hunk 23]
```diff
@@ -0,0 +1,72 @@
...
+
+    def _createRays(self):
+        origin_xs = np.random.uniform(self._limits[0][0], self._limits[0][1], self._amount)
+        origin_ys = np.random.uniform(self._limits[1][0], self._limits[1][1], self._amount)
+        origin_zs = np.random.uniform(self._limits[2][0], self._limits[2][1], self._amount)
+        direction_xs = np.random.uniform(-1, 1, self._amount)
+        direction_ys = np.random.uniform(-1, 1, self._amount)
+        direction_zs = np.random.uniform(-1, 1, self._amount)
+        for i in range(self._amount):
+            self._rays.append(Ray(Vector(origin_xs[i], origin_ys[i], origin_zs[i]), Vector(direction_xs[i], direction_ys[i], direction_zs[i])))
```
Comment (PyMarc2) [inline pytissueoptics/scene/intersection/raySource.py:72]: Not tested

File: pytissueoptics/scene/tests/tree/testSpacePartition.py [hunk 24]
```diff
@@ -14,9 +14,9 @@ def setUp(self):
         polyList = [poly, poly, poly, poly]
         bbox1 = BoundingBox(xLim=[0, 1], yLim=[0.5, 1], zLim=[0.5, 1])
         bbox2 = BoundingBox(xLim=[0.5, 1], yLim=[0.5, 1], zLim=[0.5, 1])
-        bbox3 = BoundingBox(xLim=[0.5, 1], yLim=[0.5, 1], zLim=[0.5, 1])
-        result1 = SplitNodeResult(False, "x", 1, [bbox1], [polyList])
-        result2 = SplitNodeResult(False, "x", 1, [bbox2], [polyList])
+        bbox3 = BoundingBox(xLim=[0, 0.5], yLim=[0.5, 1], zLim=[0.5, 1])
+        result1 = SplitNodeResult(False, "x", 0, [bbox1], [polyList])
+        result2 = SplitNodeResult(False, "x", 0.5, [bbox2, bbox3], [polyList])
         result3 = SplitNodeResult(True, "x", 1, [bbox3], [polyList])
```
Comment (PyMarc2) [inline pytissueoptics/scene/tests/tree/testSpacePartition.py:21]: all test for SpacePartition are failing.

File: pytissueoptics/scene/geometry/bbox.py (hunk 21 above)
Comment (PyMarc2) [reply pytissueoptics/scene/geometry/bbox.py:187]: corrected with 361e6281fe28b2a4e98c4c2d7a4c760e937bde23

File: pytissueoptics/scene/intersection/raySource.py (hunk 23 above)
Comment (PyMarc2) [reply pytissueoptics/scene/intersection/raySource.py:72]: agreed

File: pytissueoptics/scene/tests/tree/testSpacePartition.py (hunk 24 above)
Comment (PyMarc2) [reply pytissueoptics/scene/tests/tree/testSpacePartition.py]: fixed with e79604957b703583346b29b9af7add6edb409d46

**Comment by PyMarc2** [comment]
An avenue to conquer __deepcopy__, would be to make `BoundingBox` serializable. That would mean mostly reducing it to a DTO and outing the methods in a `BboxProperties` object that would be called everytime we want information on a `bbox`

```
@dataclass
BoundingBox
    limits = [[, ], [, ], [, ]]
```
thus `json.loads(json.dumps(object)` could be used and is at least 1 order of magnitude faster than __deepcopy__

**Comment by JLBegin** [comment]
Regarding the Performance section, you show better than log(n) traversal time which is impossible, but I guess you are shooting random rays so its not hitting the sphere most of the time (~99.5%)? In this case traversal time is dominated by the constant time of no-hit rays. It would be nice to truly test traversal time, so shooting relatively random rays that all collide with the sphere.

File: pytissueoptics/scene/intersection/__init__.py [hunk 25]
```diff
@@ -1,2 +1,3 @@
 from .ray import Ray
+from .raySource import RaySource, UniformRaySource, RandomPositionAndOrientationRaySource
```
Comment (JLBegin) [inline pytissueoptics/scene/intersection/__init__.py:2]: We will soon move Ray and sources to its own module

File: pytissueoptics/scene/viewer/mayavi/MayaviViewer.py [hunk 26]
```diff
@@ -35,6 +36,32 @@ def add(self, *solids: 'Solid', representation="wireframe", lineWidth=0.25, show
...
+        elif axis == "y":
+            x = [limits[0][0], limits[0][1], limits[0][1], limits[0][0]]
+            y = [value, value, value, value]
+            z = [limits[2][0], limits[2][0], limits[2][1], limits[2][1]]
+        elif axis == "z":
+            x = [limits[0][0], limits[0][1], limits[0][1], limits[0][0]]
+            y = [limits[1][0], limits[1][0], limits[1][1], limits[1][1]]
+            z = [value, value, value, value]
+        mlab.triangular_mesh(x, y, z, ((0, 1, 2), (0, 2, 3)), representation="surface", color=color, opacity=opacity)
+
```
Comment (JLBegin) [inline pytissueoptics/scene/viewer/mayavi/MayaviViewer.py]: Both are not used. I'm also not sure I see the use case. And I'm not sure to understand addPlane's signature (limits looks like a bbox but its a plane...).

**Comment by PyMarc2** [comment]
> Regarding the Performance section, you show better than log(n) traversal time which is impossible, but I guess you are shooting random rays so its not hitting the sphere most of the time (~99.5%)? In this case traversal time is dominated by the constant time of no-hit rays. It would be nice to truly test tree traversal time, so shooting relatively random rays that all collide with the sphere (ideally the rays would need a uniform origin distribution around the sphere (the whole 4pi steradians).

No I show linear, both axes are linear axes. But yes I can try with photons at the center of sphere

File: pytissueoptics/scene/intersection/__init__.py (hunk 25 above)
Comment (PyMarc2) [reply pytissueoptics/scene/intersection/__init__.py:2]: Soon

File: pytissueoptics/scene/geometry/triangle.py [hunk 27]
```diff
@@ -4,6 +4,6 @@
 
 class Triangle(Polygon):
     def __init__(self, v1: Vector, v2: Vector, v3: Vector,
-                 insideMaterial: Material = None, outsideMaterial: Material = None):
+                 insideMaterial: Material = None, outsideMaterial: Material = None, normal: Vector = None):
         super().__init__(vertices=[v1, v2, v3],
-                         insideMaterial=insideMaterial, outsideMaterial=outsideMaterial)
+                         insideMaterial=insideMaterial, outsideMaterial=outsideMaterial, normal=normal)
```
[the remark this answers is not in the document]
Comment (PyMarc2) [reply pytissueoptics/scene/geometry/triangle.py:9]: but if the normal comes from other polygons, they will always be normalized by default

File: pytissueoptics/scene/viewer/mayavi/MayaviViewer.py (hunk 26 above)
Comment (PyMarc2) [reply pytissueoptics/scene/viewer/mayavi/MayaviViewer.py]: It was used when I did plane by plane debugging
I allowed to only show partial amount of polygons from a solid and the splitting plane.

Indeed We can remove it I think.

Comment (PyMarc2) [reply pytissueoptics/scene/viewer/mayavi/MayaviViewer.py]: https://files.slack.com/files-pri/T03778D6D-F0397SA07KK/image.png

Comment (PyMarc2) [reply pytissueoptics/scene/viewer/mayavi/MayaviViewer.py]: To display this: image

Comment (PyMarc2) [reply pytissueoptics/scene/viewer/mayavi/MayaviViewer.py]: resolved with 49a93882ae0990de0c66536d4bc9bc48affc017a

File: pytissueoptics/scene/geometry/segment.py [hunk 28]
```diff
@@ -0,0 +1,9 @@
+from dataclasses import dataclass
+
+from pytissueoptics.scene.geometry import Vector
+
+
+@dataclass
+class Segment:
+    start: Vector
+    end: Vector
```
[the remark this answers is not in the document]
Comment (PyMarc2) [reply pytissueoptics/scene/geometry/segment.py]: resolved with 49a93882ae0990de0c66536d4bc9bc48affc017a

File: pytissueoptics/scene/geometry/bbox.py [hunk 29]
```diff
@@ -173,6 +177,14 @@ def shrinkTo(self, other: 'BoundingBox'):
...
             self._zLim[1] = other.zMax
 
+    def extendBy(self, value: float):
+        self._xLim[0] -= value
+        self._xLim[1] += value
+        self._yLim[0] -= value
+        self._yLim[1] += value
+        self._zLim[0] -= value
+        self._zLim[1] += value
+
```
[the remark this answers is not in the document]
Comment (PyMarc2) [reply pytissueoptics/scene/geometry/bbox.py]: resolved with 49a93882ae0990de0c66536d4bc9bc48affc017a

**Comment by JLBegin** [comment]
> > Regarding the Performance section, you show better than log(n) traversal time which is impossible, but I guess you are shooting random rays so its not hitting the sphere most of the time (~99.5%)? In this case traversal time is dominated by the constant time of no-hit rays. It would be nice to truly test tree traversal time, so shooting relatively random rays that all collide with the sphere (ideally the rays would need a uniform origin distribution around the sphere (the whole 4pi steradians).
> 
> No I show linear, both axes are linear axes. But yes I can try with photons at the center of sphere

I mean traversal time in the table, not the improvement factor. 
image

**Comment by PyMarc2** [comment]
> > > Regarding the Performance section, you show better than log(n) traversal time which is impossible, but I guess you are shooting random rays so its not hitting the sphere most of the time (~99.5%)? In this case traversal time is dominated by the constant time of no-hit rays. It would be nice to truly test tree traversal time, so shooting relatively random rays that all collide with the sphere (ideally the rays would need a uniform origin distribution around the sphere (the whole 4pi steradians).
> > 
> > 
> > No I show linear, both axes are linear axes. But yes I can try with photons at the center of sphere
> 
> I mean traversal time in the table, not the improvement factor. image

Yes I did the tests with the randomDirectionsource, positioned at the center of the sphere, here are the results:
image
image

**Comment by PyMarc2** [comment]
from my hand calculations, the time complexity of the traversal id O(3log(n)) ! Pretty good :)

**Review by JLBegin — APPROVED** [review]

**Comment by JLBegin** [comment]
Awesome.

Since the SimpleIntersectionFinder and FastIntersectionFinder both have their use cases, we will soon also consider creating a general purpose IntersectionFinder that will decide which one to use depending on the scene configuration (polygons/solid ratio). 

We will also need to move Ray, RaySource and its implementations in their own module.
